### PR TITLE
[WIP] TypeScript Parser

### PR DESCRIPTION
**Pull request by JamesHenry**
As requested by @vjeux

I have not had chance to dig into the remaining failures. As I mentioned, I am currently suffering with a bad back injury and cannot spend long on the computer at the moment.

This hacky WIP leaves us with:

```
Snapshot Summary
 › 9 snapshot tests failed in 7 test suites. Inspect your code changes or run with `npm test -- -u` to update them.

Test Suites: 9 failed, 55 passed, 64 total
Tests:       9 failed, 151 passed, 160 total
Snapshots:   9 failed, 151 passed, 160 total
Time:        9.052s
Ran all test suites.
error Command failed with exit code 1.
```

Some general notes:

- It surfaced a minor bug in `typescript-eslint-parser` for DoWhileStatements, which I have fixed, hence why the current version is the master branch of the repo. Some of the remaining issues may well be other subtle bugs in the parser which have not yet been discovered, I will be fascinated to dig in soon!

- Flow is everywhere! I was surprised to see that flowtype syntax is spread out across many parts of the test specs, and it is not possible to know from the names of the suites whether or not there will be flow inside. Was this intentional? As a result I have initially had to disable a lot of specs (see the package.json). In some case Flowtype syntax is not parsable by TypeScript, in others they produce different ASTs for the same syntax.

- There does not currently appear to be a way to run different parsers on the same files without going in and changing individual specs, so I just overwrote the flow parser usage in run_spec. Please let me know if I missed something

### CODE REVIEW

**Comment by jlongster** [comment]
I don't understand how this fixes the TypeScript problem. This looks like it's able to parse TS as an estree AST, but what about all the types? Looks like prettier won't even output TS types, and even if it does somehow, we haven't spent the time to format them nicely.

**Comment by vjeux** [comment]
@jlongster this is the first step. You should read the discussion on the typescript issue if you haven't already: https://github.com/prettier/prettier/issues/13#issuecomment-284242453

**Comment by jlongster** [comment]
I hadn't read that yet, thanks. I don't see many comments in there about supporting the TS types themselves. Are they still somewhere in the AST? What's the next step after this?

**Comment by vjeux** [comment]
The next step is to import the entire TS test suite and implement printers for all the TS-specific nodes.

**Comment by vjeux** [comment]
> Flow is everywhere! [...] Was this intentional?

We used the flow test suite to get a large corpus of crazy JavaScript edge cases and many examples are coming from the Facebook codebase which uses flow, so you just need one type to be annotated and the whole file is "flow".

This isn't intentional per se, it's more how the project came to life. It would be nice to somehow separate the many files that do not flow types in their tests.

**Comment by JamesHenry** [comment]
@jlongster Yes it is effectively the same journey we have gone on for ESLint.

Initially, for us it was all about producing a valid ESTree and ignoring the types, then slowly but surely we added more and more ESTree-like nodes in for the TypeScript type information.

Wherever there was a precedent set by flow we attempted to use that, but in some cases TypeScript has extra syntax. We should be very consistent with this behaviour, but it is currently not explicitly documented.

The TypeScript team have also requested more info about these TypeScript -> ESTree/Flow mappings that occur and I am going to document everything we do in `typescript-eslint-parser`.

I think these kind of things could be the start of some exciting converging ASTs across the different projects, perhaps TypeScript would even be able to support emitting an ESTree (+extensions) AST in the future

**Comment by JamesHenry** [comment]
Thanks @vjeux, do you have any immediately thoughts on improving/extending my current approach?

**Comment by vjeux** [comment]
@JamesHenry I would manually add

```js
run_spec(__dirname, {parser: 'typescript'});
```

for all the ones that work instead of disabling tests.

You also need to throw an exception if there are nodes that are not supported yet. This way it's very easy to figure out what doesn't work with typescript.

Once this is there, we should be able to merge this in and progressively support more of the TS nodes and fix more of the existing tests one PR at a time.

File: index.js [hunk 1]
```diff
@@ -18,9 +18,18 @@ function guessLineEnding(text) {
 }
 
 function parse(text, opts) {
-  const parseFunction = opts.parser === "flow"
-    ? parser.parseWithFlow
-    : parser.parseWithBabylon;
+  let parseFunction
+
+  switch (opts.parser) {
```
Comment (vjeux) [inline index.js]: please use if / else if ... It's a lot shorter and less error prone than switch statements.

**Comment by jlongster** [comment]
Hooking in the typescript/estree parser looks fine to me. It seems like this is a PR to get some early feedback, and not to merge.

> Once this is there, we should be able to merge this in and progressively support more of the TS nodes and fix more of the existing tests one PR at a time.

I do not want to support a `typescript` value for the parser option until we have reasonable support for it. This will just strip all the types and output JS. We should not advertise that we have support for TS when we don't really support it at all.

This code should exist on his fork until he has a reasonable pass at TS support and we can merge it all at once. This code isn't really that interesting until we get actual TS support, and will just be confusing until then.

File: package.json [hunk 2]
```diff
@@ -35,7 +35,9 @@
     "rollup-plugin-json": "2.1.0",
     "rollup-plugin-node-builtins": "2.0.0",
     "rollup-plugin-node-globals": "1.1.0",
-    "rollup-plugin-node-resolve": "2.0.0"
+    "rollup-plugin-node-resolve": "2.0.0",
+    "typescript": "2.2.1",
```
Comment (vjeux) [inline package.json]: this will need to be a "dependency" if you want people to use it in prod. But we can leave it here right now as we are experimenting with it.

**Comment by jlongster** [comment]
Also, @JamesHenry it might be worth extracting out the Flow nodes in `genericPrintNoParens` into their own file to reduce the code in `printer.js`. Then you could create another typescript file to print TS types. Would probably be easier to maintain that way.

File: src/parser.js [hunk 3]
```diff
@@ -54,4 +54,24 @@ function parseWithBabylon(text) {
...
+        range: true,
+        tokens: true,
+        attachComment: true,
+        ecmaFeatures: {
+            jsx: true,
+        }
+    })
+    return JSON.parse(JSON.stringify(result, function(key, value) {
+        if ((key === "start" || key === "end") && typeof value === "number") {
+            return undefined;
```
Comment (vjeux) [inline src/parser.js]: Why do you need to null out those?

File: package.json [hunk 4]
```diff
@@ -50,6 +52,18 @@
...
+      "tests/arrows",
+      "tests/comments",
+      "tests/flow",
+      "tests/ignore",
+      "tests/interface",
+      "tests/optional-type-name",
+      "tests/rest",
+      "tests/tabWith",
+      "tests/trailing_whitespace",
+      "tests/union_intersection",
```
Comment (jlongster) [inline package.json]: We can't disable all these tests

**Comment by JamesHenry** [comment]
Thanks guys, agree on the current usefulness. Will continue to work on it on my fork, and ping you for additional feedback as I make progress 👍

**Comment by vjeux** [comment]
> This code should exist on his fork until he has a reasonable pass at TS support and we can merge it all at once.

Good idea! I'm so used to Facebook model where we only have a single master branch and do all our work from there.

If you can maintain an "official" fork where all the ts development of prettier happens on that would be fantastic. This way more people can contribute to it and when it's good, when can merge it back inside of prettier

**Comment by jlongster** [comment]
Fwiw I'm happy to merge some super early work, but basically just adding the `typescript` parser option doesn't quite seem worth it yet :) If you get something very basic working and figure out the testing story, I'm happy to merge that with a note that TS support is experimental and we can collaborate here.

**Comment by JamesHenry** [comment]
For sure! @vjeux had just requested a super early stage WIP PR so that he could take a look

**Comment by vjeux** [comment]
I'll close this one as this is not intended to be shipped right now. I'll play with it when I have some time and @ChristopherBiscardi was interested too!

I hope you get better!

**Comment by JamesHenry** [comment]
@vjeux I'm still working on it

**Comment by JamesHenry** [comment]
Sorry for any confusion, I am keen to get this over the line, will let you know when I push the next update

**Comment by JamesHenry** [comment]
Making good progress will have an update in about an hour

**Comment by JamesHenry** [comment]
@vjeux I have just added two commits

1. Shows all the tests where the new `typescript` parser can be used as a drop in replacement for the existing parser in the test (almost always flow I think)

I did it this way first so that it was easiest for you to review. Rather than simply adding new snapshots from the typescript parser and comparing them, doing this interim replacement step allows us to be sure of the behaviour.

2. Shows all of the (annotated) failures I found.

Some are obvious:
- Flow syntax is used
- TypeScript does not support the syntax (e.g. import(), do {})

...others less so.

I believe quite strongly I have uncovered a bug in `typescript-eslint-parser` with how we convert `instanceof` operators. I think that might be the cause of 2 of the failures, but hard to be sure until I dig into that on my end.

**Comment by JamesHenry** [comment]
I must say I am chuffed with how well `typescript-eslint-parser` has done with this 😄

File: tests/es6modules/jsfmt.spec.js [hunk 5]
```diff
@@ -1,2 +1,13 @@
...
 run_spec(__dirname, { parser: 'babylon' });
+
+/**
+ * TS currently broken
+ *
+ * - export_default_function_declaration_async.js
+ *
+ * DIFF:
+ *   -export default (async function foo() {});
+ *   +export default async function foo() {}
```
Comment (vjeux) [inline tests/es6modules/jsfmt.spec.js]: This is actually a bug in both babylon and flow parsers. Looks like typescript gets it right :)

See https://github.com/prettier/prettier/issues/700

File: tests/strings/jsfmt.spec.js [hunk 6]
```diff
@@ -1,2 +1,15 @@
...
+
+/**
+ * TS currently broken
+ *
+ * - strings.js
+ *
+ * DIFF:
+ *    -  '\\uD801\\uDC28',
+ *    +
+ *    +  "\\uD801\\uDC28",
```
Comment (vjeux) [inline tests/strings/jsfmt.spec.js]: Not sure if this is it but there is a workaround in the flow parser about this. You should try to compare it against babylon and see if you match.

File: tests/unicode/jsfmt.spec.js [hunk 7]
```diff
@@ -1 +1,12 @@
 run_spec(__dirname);
+
+/**
+ * TS currently broken
+ *
+ * - keys.js
+ *
+ * DIFF:
+ *    -({ "この事はつもり素晴らしいことさ": "35jL9V" });
+ *    +({ この事はつもり素晴らしいことさ: "35jL9V" });
```
Comment (vjeux) [inline tests/unicode/jsfmt.spec.js]: This is a bug in the flow parser. You should test against babylon

File: tests/es6modules/jsfmt.spec.js (hunk 5 above)
Comment (JamesHenry) [inline tests/es6modules/jsfmt.spec.js]: So how would we move forward with this? Just record the TS as an additional snapshot?

**Comment by JamesHenry** [comment]
Thanks, @vjeux! 😄 

What would you say the next steps are for the cases where typescript is a drop-in replacement? Just add new snapshots for them?

**Comment by vjeux** [comment]
Yeah I would just add runspec({parser: 'typescript'}) for them.

I think then we can merge this PR and start doing the work of fixing everything else with a PR + test for many weeks :)

**Comment by JamesHenry** [comment]
Rebasing now as the PR https://github.com/prettier/prettier/pull/921 has changed some stuff which affects this one (including adding some unsupported syntax `::`)

**Comment by JamesHenry** [comment]
@vjeux As discussed I have added the TypeScript snapshots in the cases where they currently behave exactly like the existing snapshots do.

```
Test Suites: 440 passed, 440 total
Tests:       1373 passed, 1373 total
Snapshots:   1373 passed, 1373 total
Time:        82.054s
```

What shall we do about the annotated/commented out failures? Leave the comments in for ease of future debugging? Or strip them out of this PR for cleanliness?

**Comment by vjeux** [comment]
Wow, 82 seconds to run the entire test suite. Can you tell me how long does it run without your pull requests on your machine?

Right now on my machine it takes 18s which is fast enough so that I can run it a lot, but if it takes 1m20 this is going to harm my productivity.

**Comment by JamesHenry** [comment]
I grabbed that from the Travis build - on my machine it takes 42.079s

**Comment by JamesHenry** [comment]
Just ran master on my machine and it took 42.328s 😄

**Comment by JamesHenry** [comment]
No, I think you misunderstood the results.

Comparing like for like (my machine):

- Master branch **42.328s** (1280 snapshots)
- This PR **42.079s** (increases to 1373 snapshots) 

The result I randomly grabbed from Travis above was 82.054s. I have no idea how that compares to previous runs on Travis

**Comment by vjeux** [comment]
Ohh, I see. Sounds good then!

**Comment by JamesHenry** [comment]
> I was wondering, should we instead not run any of the typescript tests for now since we "know" they work but all of the bug fixes we are going to do going forward, then add typescript to the mix.

The only thing I would say is that rebasing this after only 1 PR went into master was already non-trivial. That PR added syntax which TypeScript doesn't support and so I had to disable it for that suite. The longer we let them drift apart the more difficult it will be to reconcile and we will effectively have to start over.

**Comment by JamesHenry** [comment]
As we mentioned before, with standard and non-standard JavaScript syntax all mixed together it is quite tricky at the moment

File: src/parser.js (hunk 3 above)
Comment (JamesHenry) [inline src/parser.js]: We don't 😄 sorry we need to do this for our tests for the parser, it's a hang up from espree being a fork of acorn. Removing!

File: tests/break-calls/jsfmt.spec.js [hunk 8]
```diff
@@ -1 +1,2 @@
 run_spec(__dirname);
+run_spec(__dirname, {parser: 'typescript'});
```
Comment (vjeux) [inline tests/break-calls/jsfmt.spec.js]: please configure your editor to add newline at the end of files

File: tests/comments/jsfmt.spec.js [hunk 9]
```diff
@@ -1,2 +1,12 @@
 run_spec(__dirname);
 run_spec(__dirname, {parser: 'babylon'});
+
+/**
```
Comment (vjeux) [inline tests/comments/jsfmt.spec.js]: please remove all those, I don't think that it's very useful to be in the code. You should keep track of them somewhere else

Comment (JamesHenry) [inline tests/comments/jsfmt.spec.js]: Sure thing!

File: tests/break-calls/jsfmt.spec.js (hunk 8 above)
Comment (JamesHenry) [inline tests/break-calls/jsfmt.spec.js]: Would you be open to an initial PR to add tooling around enforcing that? Is there a particular reason why linting or an .editorconfig has not be added up to this point?

I could also improve the consistency in file and folder naming. Right now snake, dash and camel case are all used which made putting this PR together a little more difficult.

**Comment by JamesHenry** [comment]
@christopherbiscardi Please let me know what you are confused about

**Comment by ChristopherBiscardi** [comment]
@JamesHenry oh whoops! I thought that was just a "that's unfortunate" emoji (it is unfortunate to have Flow/JS/etc mixed for this). Definitely not confused, just keeping up to date.

Please let me know if you think there's a specific area where I can be most useful. Otherwise I'll try to find work to do when I have a moment where I won't overlap with your work.

**Comment by JamesHenry** [comment]
Haha ok 😄 I agree it's not a very accurate facial expression that one!

**Comment by JamesHenry** [comment]
@vjeux I have removed all of the failing cases, so it is just about how we approach the point around newline consistency.

Totally happy to do it as a one off, or take the time to add some tooling for the project as per my comment here https://github.com/prettier/prettier/pull/915#discussion_r105010198

Just let me know how you want to do it!

File: tests/break-calls/jsfmt.spec.js (hunk 8 above)
Comment (vjeux) [inline tests/break-calls/jsfmt.spec.js]: We should probably run prettier on all those files :P

**Comment by vjeux** [comment]
Okay, so I think that it's in a good shape to be included inside of the repo. Then the next steps are going to ensure that all the TS-specific nodes are printed and that all the JS-nodes are printed correctly by going through a ton of tests that are not inside of this repo.

In order to proceed, from now on, every time something is done on the TS-part, it should be a pull request with one fix/implementation associated with one or more tests.

The only thing left before we merge it is to move typescript to devDependencies. It's a big dependency and we don't want everyone that installs prettier to have to bring it with it since it's not useable right now.

Sounds good?

**Comment by JamesHenry** [comment]
Sure thing, done!

**Comment by vjeux** [comment]
Awesome <3 Thanks a lot!

Looks like @ChristopherBiscardi you can get started adding support for TS nodes ;)

**Comment by JamesHenry** [comment]
Yes, good point, @vjeux!

@ChristopherBiscardi, I am totally happy for you to start work on that, because I will continue to work through the failures I recorded on existing tests.

Let's keep in touch on our progress 😄
